File: JavaScript/7js 3/numberJoy.js
/*
|--------------------------------------------------------------------------
| Especially Joyful Numbers
|--------------------------------------------------------------------------
|
*/

function numberJoy(n) {
   const x = String(n).split('').map(i => Number(i)).reduce((a, b) => a + b);
   const y = Number(String(x).split('').reverse().join(''));
   return x * y == n;
}


function numberJoy(n) {
   return (x = [..."" + n].reduce((a, b) => +b + a, 0)) * [..."" + x].reverse().join `` == n;
}


function numberJoy(n) {
   return (val => val * [...`${val}`].reverse().join(``) === n)
      ([...`${n}`].reduce((pre, val) => pre + +val, 0));
}


console.log(numberJoy(1997), false, 'Not a Harshad number')
console.log(numberJoy(1998), false, 'Harshad but digit sum=27, and 27x72 does not equal 1998')
console.log(numberJoy(1729), true, 'Harshad and digit sum=19, and 19x91 = 1729')
console.log(numberJoy(1458), true, "It was correct")

File: Python/7py 3/numberJoy.py
"""
|--------------------------------------------------------------------------
| Especially Joyful Numbers
|--------------------------------------------------------------------------
|
"""

def number_joy(n):
   x = sum(list(map(lambda i: int(i), str(n))))
   y = int(''.join(list(reversed(str(x)))))
   return x*y == n


def numberJoy(n):
   s=sum(int(d) for d in str(n))
   return n==s*int(str(s)[::-1])


def number_joy(n):
   s = sum(map(int,list(str(n))))
   return s*int(str(s)[::-1]) == n


print(number_joy(1997), False, 'Not a Harshad number')
print(number_joy(1998), False, 'Harshad but digit sum=27, and 27x72 does not equal 1998')
print(number_joy(1729), True, 'Harshad and digit sum=19, and 19x91 = 1729')
print(number_joy(18), False, 'Harshad but digit sum=9, and 9x9 does not equal 18')